File: src/internal/worker/events_pull.go
package worker

import (
	"context"
	"errors"
	"time"

	"github.com/ranaroussi/minigun/internal/mailgun"
	"github.com/ranaroussi/minigun/internal/store"
)

// ---------------------------------------------------------------------------
// Schedule
// ---------------------------------------------------------------------------

// BurstOffsets — relative to send.created_at, pull at these offsets.
var burstOffsetsMS = []int64{
	0,
	1 * 60 * 60 * 1000,         // +1h
	6 * 60 * 60 * 1000,         // +6h
	24 * 60 * 60 * 1000,        // +24h
}

const (
	// After the burst is exhausted, pull every DailyIntervalMS until the
	// archive window closes.
	DailyIntervalMS = 24 * 60 * 60 * 1000

	// ArchiveMaxAgeMS — total archive window, anchored to send.created_at.
	// 30 days matches Mailgun's paid-tier retention; beyond it there's
	// nothing pullable that we haven't already seen.
	ArchiveMaxAgeMS int64 = 30 * 24 * 60 * 60 * 1000

	// OverlapMS — every pull re-fetches the last 6h of events. The
	// UNIQUE(mailgun_event_id) constraint dedupes; this gives us margin
	// against Mailgun's out-of-order event arrival.
	OverlapMS int64 = 6 * 60 * 60 * 1000

	// maxPagesPerSend — hard cap on pages fetched per send per beat.
	// Prevents one cron tick from saturating after a long Mailgun outage.
	maxPagesPerSend = 50
)

// NextEventPullDueAt returns the timestamp of the next scheduled pull for
// the given send. Returns (0, true) if the send is past its archive window
// (caller should freeze).
//
// Burst phase (pulls 0..3): use the burstOffsetsMS lookup.
// Daily phase (pulls 4+):   last_pulled_at + 24h, capped by the window.
//
// Anchored to created_at (not completed_at) so the cron can dispatch
// without knowing lifecycle state.
func NextEventPullDueAt(row store.DueEventPullRow) (dueAtMs int64, frozen bool) {
	if row.EventsPullsCount < int64(len(burstOffsetsMS)) {
		return row.CreatedAtMs + burstOffsetsMS[row.EventsPullsCount], false
	}
	if !row.EventsLastPulledAtMs.Valid {
		// Defensive: pulls_count >= 4 implies a prior pull. If somehow not,
		// treat as "due now."
		return time.Now().UnixMilli(), false
	}
	next := row.EventsLastPulledAtMs.Int64 + DailyIntervalMS
	if next > row.CreatedAtMs+ArchiveMaxAgeMS {
		return 0, true
	}
	return next, false
}

// ---------------------------------------------------------------------------
// Scheduler
// ---------------------------------------------------------------------------

// RunEventsArchiveScheduler wakes every interval, finds sends with events
// due to be pulled, and ingests their events from Mailgun. Blocks until
// ctx is cancelled. No-ops when cfg.EventsArchiveEnabled is false — the
// caller (cmd/serve.go) still spawns the goroutine so flipping the env
// var doesn't require a restart of the rest of the manager.
func (m *Manager) RunEventsArchiveScheduler(ctx context.Context, interval time.Duration) {
	if interval <= 0 {
		interval = 15 * time.Minute
	}
	m.pullEventsOnce(ctx)
	t := time.NewTicker(interval)
	defer t.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
			m.pullEventsOnce(ctx)
		}
	}
}

// pullEventsOnce is the cron tick. Short-circuits when the feature flag
// is off so an operator can ship dormant in Phase 1, leave the scheduler
// running, and activate by setting EVENTS_ARCHIVE_ENABLED=true.
func (m *Manager) pullEventsOnce(ctx context.Context) {
	if !m.cfg.EventsArchiveEnabled {
		return
	}
	// Phase 5: replay any raw events whose engagement update was skipped
	// in a previous tick (insert succeeded but applyEventToEngagement
	// failed). Bounded so a backlog drains over multiple ticks.
	m.replayUnappliedEngagement(ctx, 500)
	const batchSize = 20
	nowMs := time.Now().UnixMilli()
	candidates, err := m.store.ListDueEventPulls(ctx, nowMs, ArchiveMaxAgeMS, batchSize*2)
	if err != nil {
		m.log.Warn("events-pull: list candidates", "err", err)
		return
	}
	picked := make([]store.DueEventPullRow, 0, batchSize)
	for _, c := range candidates {
		if len(picked) >= batchSize {
			break
		}
		dueAt, frozen := NextEventPullDueAt(c)
		if frozen {
			// Past the window but still wasn't frozen in the DB — run one
			// last pull to flush remaining events, after which the freeze
			// logic in RecordEventPullProgress will mark archive_complete.
			picked = append(picked, c)
			continue
		}
		if dueAt <= nowMs {
			picked = append(picked, c)
		}
	}
	for _, snd := range picked {
		select {
		case <-ctx.Done():
			return
		default:
		}
		if err := m.pullEventsForOneSend(ctx, snd); err != nil {
			m.log.Warn("events-pull: send", "send_id", snd.SendID, "err", err)
			if recordErr := m.store.RecordEventPullError(ctx, snd.SendID, err.Error()); recordErr != nil {
				m.log.Warn("events-pull: record error", "send_id", snd.SendID, "err", recordErr)
			}
		}
	}
}

// PullEventsForOneSend (exported for use by HTTP/MCP/CLI on-demand triggers)
// pulls Mailgun events for one send and updates the per-send watermark.
// Idempotent — calling it twice with no time elapsed is safe (the UNIQUE
// constraint on mailgun_event_id deduplicates).
func (m *Manager) PullEventsForOneSend(ctx context.Context, sendID string) (inserted int64, pages int, frozen bool, err error) {
	nowMs := time.Now().UnixMilli()
	candidates, err := m.store.ListDueEventPulls(ctx, nowMs, ArchiveMaxAgeMS, 1)
	if err != nil {
		return 0, 0, false, err
	}
	for _, c := range candidates {
		if c.SendID == sendID {
			return m.pullEventsForOneSendInternal(ctx, c)
		}
	}
	// The send isn't in the due-set (might be frozen, test_mode, or out
	// of window). Synthesize a minimal row so the caller still gets a
	// best-effort attempt — useful for diagnostic tools.
	listID, _ := m.store.GetSendListID(ctx, sendID)
	_ = listID
	return 0, 0, false, errors.New("send not eligible for archive (frozen / test_mode / out of window / missing domain)")
}

func (m *Manager) pullEventsForOneSend(ctx context.Context, snd store.DueEventPullRow) error {
	_, _, _, err := m.pullEventsForOneSendInternal(ctx, snd)
	return err
}

func (m *Manager) pullEventsForOneSendInternal(ctx context.Context, snd store.DueEventPullRow) (inserted int64, pages int, frozen bool, err error) {
	nowMs := time.Now().UnixMilli()
	beginMs := snd.CreatedAtMs
	if snd.EventsLastPulledThroughMs.Valid {
		beginMs = snd.EventsLastPulledThroughMs.Int64
	}
	beginMs -= OverlapMS
	endMs := nowMs

	listID, err := m.store.GetSendListID(ctx, snd.SendID)
	if err != nil {
		return 0, 0, false, err
	}

	// lastEventTs tracks the highest event_timestamp_ms we successfully
	// processed in this batch. Used by the watermark-advance logic below
	// to avoid the Phase 2 bug where hitting maxPagesPerSend with more
	// pages remaining would advance the watermark past unprocessed
	// events (silent data loss). Now: if we cap, advance only to the
	// last processed event's timestamp — the 6h overlap window on the
	// next pull will catch anything beyond it.
	var lastEventTs int64
	cappedWithMorePages := false

	var page *mailgun.EventsPage
	var pageURL string
	for pages < maxPagesPerSend {
		select {
		case <-ctx.Done():
			return inserted, pages, false, ctx.Err()
		default:
		}
		if pageURL == "" {
			page, err = m.mailgun.FetchEvents(ctx, mailgun.FetchEventsParams{
				Domain:  snd.FromDomain,
				Tag:     snd.SendID,
				BeginMs: beginMs,
				EndMs:   endMs,
				Limit:   300,
			})
		} else {
			page, err = m.mailgun.FetchEventsPage(ctx, pageURL)
		}
		if err != nil {
			return inserted, pages, false, err
		}
		pages++
		if len(page.Items) == 0 {
			break
		}
		for _, raw := range page.Items {
			ev, ok := store.NormalizeEvent(raw, snd.FromDomain, snd.SendID)
			if !ok {
				continue
			}
			if ev.EventTimestampMs > lastEventTs {
				lastEventTs = ev.EventTimestampMs
			}
			res, insertErr := m.store.InsertEventIfNew(ctx, ev)
			if insertErr != nil {
				m.log.Warn("events-pull: insert", "send_id", snd.SendID, "mailgun_event_id", ev.MailgunEventID, "err", insertErr)
				continue
			}
			if !res.Inserted {
				continue
			}
			inserted++
			// Decide whether to apply engagement. Singles without a
			// list_id, unresolved contacts, and non-engagement event
			// types all skip the apply step but immediately flip
			// engagement_applied=1 so the replay scanner doesn't keep
			// re-checking them.
			needsEngagement := listID != "" && res.ContactID.Valid && isEngagementEvent(ev.Event)
			if needsEngagement {
				if engErr := m.store.ApplyEventToEngagement(ctx, res.ContactID.String, listID, ev.Event, ev.EventTimestampMs); engErr != nil {
					m.log.Warn("events-pull: engagement upsert", "send_id", snd.SendID, "contact_id", res.ContactID.String, "err", engErr)
					// Leave engagement_applied=0; replay will retry.
					continue
				}
			}
			if markErr := m.store.MarkEngagementApplied(ctx, res.ID); markErr != nil {
				m.log.Warn("events-pull: mark engagement applied", "send_id", snd.SendID, "event_id", res.ID, "err", markErr)
			}
		}
		if page.Paging.Next == "" || page.Paging.Next == pageURL {
			break
		}
		pageURL = page.Paging.Next
		if pages >= maxPagesPerSend {
			// We're about to exit the loop with an outstanding next
			// page — flag the caller so the watermark advance is
			// conservative (Phase 5 fix for H2).
			cappedWithMorePages = true
		}
	}

	// Watermark advance strategy:
	//   * Normal exhaustion → advance to (endMs - OverlapMS). The 6h
	//     overlap guarantees the next pull re-scans for late arrivals.
	//   * Page-cap with more pages → advance only to lastEventTs (the
	//     highest event we successfully processed). The next pull will
	//     start at lastEventTs - OverlapMS and re-fetch the remaining
	//     pages via the new begin/end window — UNIQUE deduplicates the
	//     already-archived events.
	//   * Page-cap but lastEventTs is somehow 0 (unusual: cap hit with
	//     no events) → keep the existing watermark.
	newThroughMs := endMs - OverlapMS
	if cappedWithMorePages {
		if lastEventTs > 0 {
			newThroughMs = lastEventTs
		} else if snd.EventsLastPulledThroughMs.Valid {
			newThroughMs = snd.EventsLastPulledThroughMs.Int64
		} else {
			newThroughMs = snd.CreatedAtMs
		}
	}

	frozen = !cappedWithMorePages && nowMs >= snd.CreatedAtMs+ArchiveMaxAgeMS

	if progressErr := m.store.RecordEventPullProgress(ctx, snd.SendID, store.EventPullProgress{
		LastPulledAtMs:      nowMs,
		LastPulledThroughMs: newThroughMs,
		Inserted:            inserted,
		Freeze:              frozen,
	}); progressErr != nil {
		return inserted, pages, frozen, progressErr
	}
	return inserted, pages, frozen, nil
}

// isEngagementEvent reports whether the event type drives an update to
// the contact_engagement summary. Used to decide whether to mark
// engagement_applied=1 directly (non-engagement event) vs. wait for the
// apply step to succeed first.
func isEngagementEvent(eventType string) bool {
	switch eventType {
	case "delivered", "opened", "clicked":
		return true
	default:
		return false
	}
}

// replayUnappliedEngagement scans for raw events whose engagement update
// never landed (insert succeeded but apply failed in a prior tick). For
// each it looks up the list_id and re-runs ApplyEventToEngagement, then
// flips engagement_applied=1. Bounded per tick so a backlog drains
// gracefully without saturating the cron.
//
// Idempotency: ApplyEventToEngagement's UPSERT semantics with the MAX()
// guards make double-application safe for last_*_at_ms columns, but
// total_* counters would double-count if naively replayed. The flag
// is what prevents that — we only replay events explicitly marked 0,
// and we mark them 1 immediately after the apply call returns nil.
func (m *Manager) replayUnappliedEngagement(ctx context.Context, limit int) {
	events, err := m.store.ListUnappliedEngagementEvents(ctx, limit)
	if err != nil {
		m.log.Warn("events-pull: replay list", "err", err)
		return
	}
	if len(events) == 0 {
		return
	}
	for _, ev := range events {
		select {
		case <-ctx.Done():
			return
		default:
		}
		if !ev.ContactID.Valid || !isEngagementEvent(ev.Event) {
			if markErr := m.store.MarkEngagementApplied(ctx, ev.ID); markErr != nil {
				m.log.Warn("events-pull: replay mark", "event_id", ev.ID, "err", markErr)
			}
			continue
		}
		listID, lookupErr := m.store.GetSendListID(ctx, ev.MgSendID)
		if lookupErr != nil {
			m.log.Warn("events-pull: replay list_id lookup", "event_id", ev.ID, "err", lookupErr)
			continue
		}
		if listID == "" {
			if markErr := m.store.MarkEngagementApplied(ctx, ev.ID); markErr != nil {
				m.log.Warn("events-pull: replay mark", "event_id", ev.ID, "err", markErr)
			}
			continue
		}
		if engErr := m.store.ApplyEventToEngagement(ctx, ev.ContactID.String, listID, ev.Event, ev.EventTimestampMs); engErr != nil {
			m.log.Warn("events-pull: replay apply", "event_id", ev.ID, "err", engErr)
			continue
		}
		if markErr := m.store.MarkEngagementApplied(ctx, ev.ID); markErr != nil {
			m.log.Warn("events-pull: replay mark", "event_id", ev.ID, "err", markErr)
		}
	}
}

File: worker/src/send/events_pull.ts
import { Env, engagementStatsEnabled } from '../env';
import { fetchEvents, MailgunAPIError } from '../lib/mailgun';
import {
  DueEventPullRow,
  checkpointEventPullThrough,
  clickStmt,
  engagementStmt,
  getSendListID,
  listDueEventPulls,
  lookupContactIDsByEmails,
  messageEngagementStmt,
  normalizeEvent,
  recordEventPullError,
  recordEventPullProgress,
} from '../store/events';
import { hasActiveSend } from '../store/sends';

// ---------------------------------------------------------------------------
// Schedule
// ---------------------------------------------------------------------------

// Burst beats: relative to send.created_at, fire at these offsets.
const BURST_OFFSETS_MS = [
  0,
  1 * 60 * 60 * 1000,        // +1h
  6 * 60 * 60 * 1000,        // +6h
  24 * 60 * 60 * 1000,       // +24h
];

// After the burst is exhausted, pull every DAILY_INTERVAL_MS until the
// archive window closes.
const DAILY_INTERVAL_MS = 24 * 60 * 60 * 1000;

// Total archive window, anchored to send.created_at. After this elapses
// the send is frozen — no more polls, no more reads against Mailgun.
// 14 days: opens/clicks are overwhelmingly front-loaded (first few days),
// so the second half of Mailgun's 30-day retention folds almost nothing
// while keeping the send in the daily-pull candidate set and writing a
// checkpoint each beat. Freezing at 14d trims that write/read tail with
// negligible engagement loss. Raise back toward 30d if long-tail opens
// ever matter more than the D1 budget.
export const ARCHIVE_MAX_AGE_MS = 14 * 24 * 60 * 60 * 1000;

// Hard cap on pages fetched per send per beat. Each page is one Mailgun
// fetch + one batched contact lookup + chunked D1 write batches, so this
// bounds the CPU/subrequest budget a single cron tick spends on one send
// (300 events/page × 16 pages = 4.8k events/beat). A send with a larger
// backlog is drained across consecutive cron ticks: while capped, the
// beat checkpoints its watermark WITHOUT bumping events_pulls_count, so
// nextDueAt keeps it immediately due until it catches up. The per-page
// checkpoint means a CPU-killed beat still keeps the pages it finished —
// only the dense same-second delivery burst right after a send risks
// hitting the limit at this cap, and that section is idempotent.
const MAX_PAGES_PER_BEAT = 16;

// Max statements per D1 batch() call. A page of 300 events can emit up to
// ~3 writes each; we flush in chunks so no single batch is oversized.
const BATCH_CHUNK = 90;

// Returns the timestamp of the next scheduled pull beat for `send`, or
// null if the send is past the archive window (caller should freeze).
//
// Burst phase (pulls 0..3): use the BURST_OFFSETS_MS lookup.
// Daily phase (pulls 4+):   last_pulled_at + 24h, capped by the window.
//
// Anchoring everything to created_at (not completed_at) means the cron
// can dispatch without knowing the send's lifecycle status. For typical
// sends (complete in minutes), +1h-after-created ≈ +1h-after-completed.
// For long-running sends (slow re-warms over weeks), pulling during
// the send gives operators real-time deliverability monitoring.
export function nextDueAt(row: {
  created_at_ms: number;
  events_pulls_count: number;
  events_last_pulled_at_ms: number | null;
}): number | null {
  if (row.events_pulls_count < BURST_OFFSETS_MS.length) {
    return row.created_at_ms + BURST_OFFSETS_MS[row.events_pulls_count]!;
  }
  if (row.events_last_pulled_at_ms === null) {
    // Shouldn't happen — pulls_count >= 4 implies we've pulled — but be
    // defensive and treat this as "due now."
    return Date.now();
  }
  const next = row.events_last_pulled_at_ms + DAILY_INTERVAL_MS;
  if (next > row.created_at_ms + ARCHIVE_MAX_AGE_MS) return null;
  return next;
}

// ---------------------------------------------------------------------------
// Orchestration
// ---------------------------------------------------------------------------

// Cron entrypoint. Called from the worker's scheduled handler. No-ops
// when ENGAGEMENT_STATS_ENABLED is unset (engagement retrieval feature flag).
export async function pullDueSendEvents(env: Env, limit = 20): Promise<void> {
  if (!engagementStatsEnabled(env)) return;
  // Pause retrieval while a send is in flight. The events-pull can burn the
  // whole scheduled-handler CPU budget folding the dense delivery burst of a
  // large send, which starves sweepStuckSends and wedges the very send that
  // is generating those events. Deferring the pull by a tick (or a few, for
  // long sends) is harmless — the burst/daily schedule is anchored to
  // created_at and stays "due" until it runs, so nothing is dropped.
  try {
    if (await hasActiveSend(env.DB)) return;
  } catch (err) {
    console.error('events-pull: active-send check', err);
    return;
  }
  const now = Date.now();
  let candidates: DueEventPullRow[];
  try {
    candidates = await listDueEventPulls(env.DB, now, ARCHIVE_MAX_AGE_MS, limit * 2);
  } catch (err) {
    console.error('events-pull: list candidates', err);
    return;
  }
  // Filter through nextDueAt (the SQL pulled "non-frozen, within window"
  // but the burst-vs-daily phase logic is in JS so the index can stay
  // simple). Take up to `limit` candidates whose due time has arrived.
  const picked: DueEventPullRow[] = [];
  for (const c of candidates) {
    if (picked.length >= limit) break;
    const due = nextDueAt(c);
    if (due === null) {
      // Send is past the archive window but still wasn't marked complete
      // (could happen if it never had completed_at set). Run one final
      // pull to flush remaining events, then it'll freeze.
      picked.push(c);
      continue;
    }
    if (due <= now) picked.push(c);
  }
  for (const send of picked) {
    try {
      await pullEventsForOneSend(env, send);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      console.error('events-pull: send', send.send_id, msg);
      try {
        await recordEventPullError(env.DB, send.send_id, msg);
      } catch (innerErr) {
        console.error('events-pull: record error', send.send_id, innerErr);
      }
    }
  }
}

// Pull events for one send. Pages through Mailgun's events API until
// either the cursor exhausts or MAX_PAGES_PER_SEND is hit. Each event is
// normalized, its recipient resolved to a contact_id, and folded into
// the two engagement rollups (contact_message_engagement always;
// contact_engagement when the send is list-tied).
//
// The window begins strictly AFTER the previous pull's highest event
// timestamp (default: the send's created_at on the first pull). Because
// Mailgun returns events in time order and we advance the watermark to
// the last event we saw, the next pull never re-fetches an
// already-counted event — so the per-call counter increments stay
// correct without any dedup ledger.
export async function pullEventsForOneSend(
  env: Env,
  send: DueEventPullRow,
): Promise<{ inserted: number; pages: number; frozen: boolean }> {
  const nowMs = Date.now();
  // +1ms makes the lower bound exclusive at storage (ms) granularity.
  //
  // Known tradeoffs of the no-ledger design (accepted, not bugs):
  //  1. Counter drift on crash/retry. Events are folded into the rollups
  //     before the watermark is persisted below, and the counter UPSERTs
  //     (total_opens+1, messages_since_last_engagement+1, ...) are not
  //     idempotent. If this run dies mid-pull or recordEventPullProgress
  //     fails, the next beat re-applies the events already folded,
  //     inflating the *counts*. Timestamp fields converge regardless
  //     (MIN/MAX), so only integer counters drift; recency-based hygiene
  //     is unaffected.
  //  2. Same-millisecond skip at the exclusive boundary. The next pull
  //     resumes at lastEventTs+1ms, so any event sharing that exact ms
  //     not seen this run — a late, eventually-consistent arrival, or one
  //     stranded past the MAX_PAGES_PER_SEND cap — is skipped permanently.
  //     Bounded-probability given ms granularity and the 15k-events/send
  //     cap; the cost of holding no cursor.
  const beginMs =
    send.events_last_pulled_through_ms !== null
      ? send.events_last_pulled_through_ms + 1
      : send.created_at_ms;
  const endMs = nowMs;

  // One list_id lookup per send (cached locally). Singles without a
  // list_id return null and skip the contact_engagement UPSERT.
  const listID = await getSendListID(env.DB, send.send_id);

  let inserted = 0;
  let pages = 0;
  let pageURL: string | undefined;
  // Highest event timestamp durably checkpointed so far. The next pull
  // resumes strictly after it. Advanced per-page (not just at the end) so
  // a CPU-killed beat still makes forward progress.
  let lastEventTs = lastThroughOnEntry(send);
  let cappedWithMorePages = false;

  while (pages < MAX_PAGES_PER_BEAT) {
    const page = pageURL
      ? await fetchEvents(env, { pageURL })
      : await fetchEvents(env, {
          domain: send.from_domain,
          tag: send.send_id,
          beginMs,
          endMs,
          limit: 300,
        });
    pages++;

    if (!page.items || page.items.length === 0) break;

    // Normalize the page, then resolve every recipient in ONE read
    // instead of a serial lookup per event (the dominant CPU cost on
    // large sends).
    const events = page.items
      .map(normalizeEvent)
      .filter((e): e is NonNullable<typeof e> => e !== null);
    const contactByEmail = await lookupContactIDsByEmails(
      env.DB,
      events.map((e) => e.recipient),
    );

    // Build all writes for this page, then flush them in ordered batches.
    // Order is preserved across chunks so the non-idempotent counter
    // UPSERTs (total_opens + 1, ...) accumulate correctly.
    const stmts: D1PreparedStatement[] = [];
    let pageMaxTs = 0;
    for (const ev of events) {
      if (ev.event_timestamp_ms > pageMaxTs) pageMaxTs = ev.event_timestamp_ms;
      const contactID = contactByEmail.get(ev.recipient);
      if (!contactID) continue; // event for an unknown address

      inserted++;
      if (isMessageEvent(ev.event)) {
        const s = messageEngagementStmt(
          env.DB,
          send.send_id,
          contactID,
          listID,
          ev.event,
          Math.floor(ev.event_timestamp_ms / 1000),
          ev.severity,
          ev.reason,
        );
        if (s) stmts.push(s);
      }
      if (listID !== null && isEngagementEvent(ev.event)) {
        const s = engagementStmt(env.DB, contactID, listID, ev.event, ev.event_timestamp_ms);
        if (s) stmts.push(s);
      }
      if (ev.event === 'clicked' && ev.url) {
        const s = clickStmt(
          env.DB,
          send.send_id,
          contactID,
          listID,
          ev.url,
          Math.floor(ev.event_timestamp_ms / 1000),
        );
        if (s) stmts.push(s);
      }
    }

    for (let i = 0; i < stmts.length; i += BATCH_CHUNK) {
      await env.DB.batch(stmts.slice(i, i + BATCH_CHUNK));
    }

    // Durable per-page checkpoint: advance the watermark to this page's
    // max event timestamp WITHOUT bumping events_pulls_count, so a beat
    // that dies on the next page resumes after this one.
    if (pageMaxTs > lastEventTs) {
      lastEventTs = pageMaxTs;
      await checkpointEventPullThrough(env.DB, send.send_id, {
        last_pulled_through_ms: lastEventTs,
        inserted,
      });
      inserted = 0; // already folded into events_archive_count above
    }

    // Follow Mailgun's pagination cursor. When items is empty OR the
    // next URL is missing/identical, we've reached the end.
    const next = page.paging?.next;
    if (!next || next === pageURL) break;
    pageURL = next;
    if (pages >= MAX_PAGES_PER_BEAT) {
      // Exiting with an outstanding next page — this scheduled pull
      // hasn't caught up. Don't bump the burst counter; the next cron
      // tick continues paging from the checkpointed watermark.
      cappedWithMorePages = true;
    }
  }

  // While still draining a backlog (capped with more pages), the per-page
  // checkpoint already persisted forward progress and we deliberately do
  // NOT advance the burst schedule. Only when this scheduled pull has
  // caught up do we bump events_pulls_count and (if past the window)
  // freeze the send.
  if (!cappedWithMorePages) {
    const newThroughMs = lastEventTs > 0 ? lastEventTs : beginMs - 1;
    const frozen = nowMs >= send.created_at_ms + ARCHIVE_MAX_AGE_MS;
    await recordEventPullProgress(env.DB, send.send_id, {
      last_pulled_at_ms: nowMs,
      last_pulled_through_ms: newThroughMs,
      inserted,
      freeze: frozen,
    });
    return { inserted, pages, frozen };
  }

  return { inserted, pages, frozen: false };
}

// The watermark already covered on entry — events strictly after it are
// what this beat fetches. Used to seed lastEventTs so an empty/no-new
// page doesn't rewind the checkpoint.
function lastThroughOnEntry(send: DueEventPullRow): number {
  return send.events_last_pulled_through_ms !== null
    ? send.events_last_pulled_through_ms
    : 0;
}

// isEngagementEvent reports whether the event type drives an update to
// the per-(contact, list) contact_engagement summary.
function isEngagementEvent(eventType: string): boolean {
  return eventType === 'delivered' || eventType === 'opened' || eventType === 'clicked';
}

// isMessageEvent reports whether the event type updates the per-(send,
// contact) message engagement row. Broader than isEngagementEvent (the
// per-list rollup gate): cme also records acceptance, failure, complaint,
// and unsubscribe so a single message's full lifecycle is queryable.
function isMessageEvent(eventType: string): boolean {
  switch (eventType) {
    case 'accepted':
    case 'delivered':
    case 'opened':
    case 'clicked':
    case 'failed':
    case 'rejected':
    case 'complained':
    case 'unsubscribed':
      return true;
    default:
      return false;
  }
}

// Surface for direct on-demand invocation (POST /mailgun/events/pull?send_id=).
// Phase 3 will wire this to an HTTP route; exposing it here for now keeps
// the dependency direction clean.
export { MailgunAPIError };
